### PR TITLE
Minor features for 3.0

### LINKED ISSUE #308: matmul with 3+ arguments missing
ArrayFire 2.1 had two matrix multiplication functions that took three and four `af::array`s as arguments. This is missing from 3.0.

### LINKED ISSUE #347: Support Convolve using Frequency domain at large sizes.
Right now CUDA and OpenCL have limitation on kernel sizes. Remove that limitation and use frequency domain for convolution.

### LINKED ISSUE #589: Make convolve switch between spatial and fft based convolutions
- Add an `enum` at the end that enables the user to choose the type of convolution manually. 

The three different types of convolution:

- Spatial
- Frequency
- Automatic

The default should be "auto".

### PR DESCRIPTION
**Pull request by pavanky**
- Fixes #308 
- Fixes #347 
- Fixes #589

### CODE REVIEW

File: include/af/defines.h [hunk 1]
```diff
@@ -140,18 +151,18 @@ typedef enum {
...
-    AF_SOLVE_TRANS     = 512,  ///< Solve A.T() (non-conjugate transpose)
-    AF_SOLVE_UPPERTRI  = 1024, ///< Solve uppertri(A) (upper triangular system)
-    AF_SOLVE_LOWERTRI  = 2048, ///< Solve lowertri(A) (lower triangular system)
-    AF_SOLVE_TRIDIAG   = 4096,
-    AF_SOLVE_BLKDIAG   = 8192
-} af_solve_t;
+    AF_MAT_NONE       = 0,    ///< Default
+    AF_MAT_TRANS      = 1,    ///< Data needs to be transposed
+    AF_MAT_CTRANS     = 2,    ///< Data needs to be conjugate tansposed
+    AF_MAT_UPPER      = 32,   ///< Matrix is upper triangular
```
Comment (umar456) [inline include/af/defines.h]: Why are there large jumps in the flag values?

**Comment by shehzan10** [comment]
You didn't change `af_blas_transpose` enum values to match those of blas?

File: include/af/defines.h (hunk 1 above)
Comment (pavanky) [inline include/af/defines.h]: Left in case we need other things in the future.

**Comment by pavanky** [comment]
@shehzan10 There is no consistent enum values for "blas". CBLAS is different from cuBLAS which is different from clBLAS.

**Pull request merged**